core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * get_pointer searches the pointer in the registry by string
+ */
+template <typename T>
+inline std::shared_ptr<T> get_pointer(const pnode& config,
+                                      const registry& context,
+                                      type_descriptor td);
+
+
+/**
+ * get_factory builds the factory from config (map) or searches the pointers in
+ * the registry by string.
+ */
+template <typename T>
+deferred_factory_parameter<T> get_factory(const pnode& config,
+                                          const registry& context,
+                                          type_descriptor td);
+
+/**
+ * specialize for const LinOpFactory
+ */
+template <>
+deferred_factory_parameter<const LinOpFactory> get_factory<const LinOpFactory>(
+    const pnode& config, const registry& context, type_descriptor td);
+
+/**
+ * specialize for const stop::CriterionFactory
+ */
+template <>
+deferred_factory_parameter<const stop::CriterionFactory>
+get_factory<const stop::CriterionFactory>(const pnode& config,
+                                          const registry& context,
+                                          type_descriptor td);
+
+/**
+ * get_factory_vector will gives a vector of factory by calling get_factory.
+ */
+template <typename T>
+inline std::vector<deferred_factory_parameter<T>> get_factory_vector(
+    const pnode& config, const registry& context, type_descriptor td);
+
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for integral type
+ */
+template <typename IndexType>
+inline
+    typename std::enable_if<std::is_integral<IndexType>::value, IndexType>::type
+    get_value(const pnode& config);
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for floating point type
+ */
+template <typename ValueType>
+inline typename std::enable_if<std::is_floating_point<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config);
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for complex type
+ */
+template <typename ValueType>
+inline typename std::enable_if<gko::is_complex_s<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config);
+
+
+// type_string providing the mapping from type to string.
+template <typename T>
+struct type_string {};
+
+#define TYPE_STRING_OVERLOAD(_type, _str)         \
+    template <>                                   \
+    struct type_string<_type> {                   \
+        static std::string str() { return _str; } \
+    }
+
+TYPE_STRING_OVERLOAD(void, "void");
+TYPE_STRING_OVERLOAD(double, "double");
+TYPE_STRING_OVERLOAD(float, "float");
+TYPE_STRING_OVERLOAD(std::complex<double>, "complex<double>");
+TYPE_STRING_OVERLOAD(std::complex<float>, "complex<float>");
+TYPE_STRING_OVERLOAD(int32, "int");
```

Comment:
IMO we should use the fixed size types:
```suggestion
TYPE_STRING_OVERLOAD(double, "float64");
TYPE_STRING_OVERLOAD(float, "float32");
TYPE_STRING_OVERLOAD(std::complex<double>, "complex<float64>");
TYPE_STRING_OVERLOAD(std::complex<float>, "complex<float32>");
TYPE_STRING_OVERLOAD(int32, "int32");
```